### PR TITLE
Product management

### PR DESCRIPTION
**Pull request by arijit1111**

### CODE REVIEW

File: Modules/Product/Http/Controllers/ProductController.php [hunk 1]
```diff
@@ -0,0 +1,1193 @@
...
+        $color_key = in_array('Color', $option_types) ? array_search("Color", $option_types) + 1 : 0;
+        $colors = array();
+        $swatches = array();
+        $color_options = json_decode($request->input('colorOptionItems'), true);
+        if (is_countable($color_options) && count($color_options) > 0) {
+            foreach ($color_options as $color) {
+                if (!in_array($color['name'], $colors)) {
+                    $colors[] = $color['name'];
+                    $swatch_image = $color["img"];
+                    if (isset($swatch_image) && $swatch_image != "" && filter_var($swatch_image, FILTER_VALIDATE_URL) === false) {
```
Comment (shoaibrehman) [inline Modules/Product/Http/Controllers/ProductController.php]: Replace ` $swatch_image != ""` with `!empty(swatch_image)`

File: Modules/Product/Http/Controllers/ProductController.php [hunk 2]
```diff
@@ -0,0 +1,1193 @@
...
+                    $query->where('status', 'publish');
+                    break;
+                case 'unpublish':
+                    $query->where('status', 'unpublish');
+                    break;
+                default:
+                    break;
+            }
+            switch ($request->sort_key) {
+                case 1:
```
Comment (shoaibrehman) [inline Modules/Product/Http/Controllers/ProductController.php]: `case 1:` has the same code block as `default:`. Please remove `case 1:`

File: Modules/Product/Http/Controllers/ProductController.php [hunk 3]
```diff
@@ -0,0 +1,1193 @@
+<?php
+
+namespace Modules\Product\Http\Controllers;
+
+use Illuminate\Contracts\Support\Renderable;
```
Comment (shoaibrehman) [inline Modules/Product/Http/Controllers/ProductController.php]: `use Illuminate\Contracts\Support\Renderable;` is not used anywhere in the file. Please remove.

File: Modules/Product/Http/Controllers/ProductController.php [hunk 4]
```diff
@@ -0,0 +1,1193 @@
...
+{
+    public function index()
+    {
+        return view('product::index');
+    }
+
+    public function create(Request $request)
+    {
+        $vndr_upload_path = '/uploads/products/';
+        $folderPath = public_path() . $vndr_upload_path;
```
Comment (shoaibrehman) [inline Modules/Product/Http/Controllers/ProductController.php]: `$folderPath` & `product_images` aren't used anywhere. Please remove.

File: Modules/Product/Http/Controllers/ProductController.php [hunk 5]
```diff
@@ -0,0 +1,1193 @@
...
+        $response = ['res' => true, 'msg' => "Added Successfully", 'data' => ""];
+
+        return response()->json($response);
+    }
+
+    public function fetchProductBySort(Request $request)
+    {
+        $redis = Redis::connection();
+        $result_array = array();
+        $redis = Redis::connection();
```
Comment (shoaibrehman) [inline Modules/Product/Http/Controllers/ProductController.php]: Duplicate declaration. You have already defined `$redis` on line 293.

File: Modules/Product/Http/Controllers/ProductController.php (hunk 1 above)
Comment (arijit1111) [reply Modules/Product/Http/Controllers/ProductController.php]: Done

File: Modules/Product/Http/Controllers/ProductController.php (hunk 2 above)
Comment (arijit1111) [reply Modules/Product/Http/Controllers/ProductController.php]: Done

File: Modules/Product/Http/Controllers/ProductController.php (hunk 3 above)
Comment (arijit1111) [reply Modules/Product/Http/Controllers/ProductController.php]: Done

File: Modules/Product/Http/Controllers/ProductController.php (hunk 4 above)
Comment (arijit1111) [reply Modules/Product/Http/Controllers/ProductController.php]: Done

File: Modules/Product/Http/Controllers/ProductController.php (hunk 5 above)
Comment (arijit1111) [reply Modules/Product/Http/Controllers/ProductController.php]: Done

File: Modules/Product/Http/Controllers/ProductController.php [hunk 6]
```diff
@@ -0,0 +1,1821 @@
...
+                    break;
+            }
+            switch ($request->sort_key) {
+                case 2:
+                    $query->orderBy('name', 'DESC');
+                    break;
+                case 3:
+                    $query->orderBy('updated_at', 'DESC');
+                    $query->orderBy('id', 'DESC');
+                    //$query->orderByRaw("TIME(updated_at) DESC");
```
Comment (shoaibrehman) [inline Modules/Product/Http/Controllers/ProductController.php]: Remove commented code.

File: Modules/Product/Http/Controllers/ProductController.php [hunk 7]
```diff
@@ -0,0 +1,1821 @@
...
+        $redis = Redis::connection();
+
+        $search = $request->search_key && !in_array($request->search_key, array('undefined', 'null')) ? $request->search_key : '';
+        $existredis = Redis::exists("brandproduct:FetchProducts:" . $request->page . ":" . $search . ":" . $request->user_id . ":" . $request->status);
+        if ($existredis > 0) {
+            $cachedproducts = Redis::get("brandproduct:FetchProducts:" . $request->page . ":" . $search . ":" . $request->user_id . ":" . $request->status);
+            $allfetchproduct = json_decode($cachedproducts, false);
+            $response = ['res' => true, 'msg' => "", 'data' => $allfetchproduct];
+        } else {
+            $product_variations_tbl = DB::raw("(SELECT product_id,id as variant_id,value1,value2,value3,sku as vsku,stock as vstock,image as vimage
```
Comment (shoaibrehman) [inline Modules/Product/Http/Controllers/ProductController.php:373]: Raw DB query needs to be replaced.

File: Modules/Product/Http/Controllers/ProductController.php [hunk 8]
```diff
@@ -0,0 +1,1821 @@
...
+                $response = ['res' => true, 'msg' => "Imported Successfully", 'data' => ""];
+            } else {
+                $response = ['res' => false, 'msg' => "Enter valid information", 'data' => ""];
+            }
+        }
+
+        return response()->json($response);
+    }
+
+    public function ImportShopify(Request $request)
```
Comment (shoaibrehman) [inline Modules/Product/Http/Controllers/ProductController.php]: Shopify functionality should be created in its own module to make the functionality modular. This is not the right approach.

**Review by shoaibrehman — CHANGES_REQUESTED** [review]

**Pull request closed without merging**